mysql-test/suite/rpl/t/percona_rpl_mts_backup_lock_ps3345.test

```diff
@@ -0,0 +1,65 @@
+################################################################################
+# PS-3345: LP #1527463: Waiting for binlog lock
+#
+# 3-way deadlock can be achieved with 2 slave threads working and parallel and
+# with 1 slave client that executes LOCK BINLOG FOR BACKUP. 
+#
+# And the deadlock is:
+# worker0: applying INSERT INTO t1 VALUES(11, NULL);
+# worker1: applying INSERT INTO t1 VALUES(12, NULL);
+# worker1: calls backup_binlog_lock.acquire_protection()
+# worker1: waits for worker0 in wait_for_its_turn()
+# client: executes LOCK BINLOG FOR BACKUP
+# client: waits in backup_binlog_lock.acquire(), but protection is acquired by worker1
+# worker0: calls backup_binlog_lock.acquire_protection(), but it's blocked by client
+#
+# This test verifies the deadlock can be found and handled correctly.
+# When it finds the deadlock, it throws assert.
+################################################################################
+
+--source include/have_debug.inc
+--source include/master-slave.inc
+
+--source include/rpl_connection_slave.inc
+--source include/only_mts_slave_parallel_workers.inc
+--source include/only_mts_slave_parallel_type_logical_clock.inc
+--source include/stop_slave_sql.inc
+
+--echo #
+--echo # Case 1: 3-way deadlock with 2 slave threads and 1 slave client 
+--echo #
+--source include/rpl_connection_master.inc
+CREATE TABLE t1(c1 INT PRIMARY KEY, c2 INT, INDEX(c2)) ENGINE = InnoDB;
+SET debug = "+d,set_commit_parent_100";
+INSERT INTO t1 VALUES(11, NULL);
+INSERT INTO t1 VALUES(12, NULL);
+
+--source include/rpl_connection_slave1.inc
+SET GLOBAL debug = "+d,delay_slave_worker_0";
+SET DEBUG_SYNC = 'before_lock_binlog_for_backup SIGNAL signal_lock_binlog_for_backup';
+--source include/start_slave_sql.inc
+--send LOCK BINLOG FOR BACKUP
+
+--source include/rpl_connection_slave.inc
+--let $assert_text= Check how many threads are waiting for binlog lock
+--let $assert_cond= [ SELECT COUNT(*) FROM INFORMATION_SCHEMA.PROCESSLIST WHERE STATE = "Waiting for binlog lock" ] = 1
+--source include/assert.inc
```

Comment:
There should be some synchronisation point before we check for this condition, otherwise test would be unstable. It actually failed for me with --repeat=10

```
...
[ 30%] rpl.percona_rpl_mts_backup_lock_ps3345 'mix' [ pass ]    571
[ 33%] rpl.percona_rpl_mts_backup_lock_ps3345 'mix' [ pass ]    726
[ 36%] rpl.percona_rpl_mts_backup_lock_ps3345 'row' [ pass ]    626
[ 40%] rpl.percona_rpl_mts_backup_lock_ps3345 'row' [ pass ]    916
[ 43%] rpl.percona_rpl_mts_backup_lock_ps3345 'row' [ pass ]    548
[ 46%] rpl.percona_rpl_mts_backup_lock_ps3345 'row' [ pass ]    578
[ 50%] rpl.percona_rpl_mts_backup_lock_ps3345 'row' [ pass ]    629
[ 53%] rpl.percona_rpl_mts_backup_lock_ps3345 'row' [ pass ]    603
[ 56%] rpl.percona_rpl_mts_backup_lock_ps3345 'row' [ fail ]
        Test ended at 2019-09-16 00:36:19

CURRENT_TEST: rpl.percona_rpl_mts_backup_lock_ps3345
mysqltest: In included file ./include/assert.inc at line 81:
included from /home/sergei/git/percona-server/mysql-test/suite/rpl/t/percona_rpl_mts_backup_lock_ps3345.test at line 47:
At line 80: Test assertion failed in assert.inc

The result from queries just before the failure was:

...
```

Reply:
During deadlock we have 2 or 3 threads at "Waiting for binlog lock".
Without deadlock it's 0 or 1 so I changed the condition to:
```[ SELECT COUNT(*) FROM INFORMATION_SCHEMA.PROCESSLIST WHERE STATE = "Waiting for binlog lock" ] < 2```